index.js

```diff
@@ -2127,6 +2077,62 @@
                    foldable);
   }
 
+  //# takeWhile :: (Applicative f, Foldable f, Monoid (f a)) => (a -> Boolean, f a) -> f a
+  //.
+  //. Discards the first element which does not satisfy the predicate, and all
+  //. subsequent elements.
+  //.
+  //. This function is derived from [`concat`](#concat), [`empty`](#empty),
+  //. [`of`](#of), and [`reduce`](#reduce).
+  //.
+  //. See also [`dropWhile`](#dropWhile).
+  //.
+  //. ```javascript
+  //. > takeWhile (s => /x/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xy', 'xz', 'yx']
+  //.
+  //. > takeWhile (s => /y/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xy']
+  //.
+  //. > takeWhile (s => /z/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. []
+  //. ```
+  function takeWhile(pred, foldable) {
+    var F = foldable.constructor;
+    return (reduce (function(acc, x) {
+      var take = acc.take && pred (x);
+      return {take: take, xs: take ? concat (acc.xs, of (F, x)) : acc.xs};
+    }, {take: true, xs: empty (F)}, foldable)).xs;
+  }
+
+  //# dropWhile :: (Applicative f, Foldable f, Monoid (f a)) => (a -> Boolean, f a) -> f a
+  //.
+  //. Retains the first element which does not satisfy the predicate, and all
+  //. subsequent elements.
+  //.
+  //. This function is derived from [`concat`](#concat), [`empty`](#empty),
+  //. [`of`](#of), and [`reduce`](#reduce).
+  //.
+  //. See also [`takeWhile`](#takeWhile).
+  //.
+  //. ```javascript
+  //. > dropWhile (s => /x/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['yz', 'zx', 'zy']
+  //.
+  //. > dropWhile (s => /y/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xz', 'yx', 'yz', 'zx', 'zy']
+  //.
+  //. > dropWhile (s => /z/.test (s), ['xy', 'xz', 'yx', 'yz', 'zx', 'zy'])
+  //. ['xy', 'xz', 'yx', 'yz', 'zx', 'zy']
+  //. ```
+  function dropWhile(pred, foldable) {
+    var F = foldable.constructor;
+    return (reduce (function(acc, x) {
+      var drop = acc.drop && pred (x);
+      return {drop: drop, xs: drop ? acc.xs : concat (acc.xs, of (F, x))};
```

Comment:
This is cosmetic only, but are we allowed to use ES6 features like:

```
const drop = acc.drop && pred (x);
const xs = drop ? acc.xs : concat (acc.xs, of (F, x))

return { drop, xs }
```

?

Reply:
All Sanctuary projects target ECMAScript 5.1, which I believe is the best version of the language that will ever exist.

I'm not trolling. :)

Reply:
I have to get your reasoning on that at some point, but I guess that is a discussion for a different time. :smile:

Reply:
I believe the proportion of well-considered language features whose virtues justify their complexity costs is decreasing as the language evolves. I believe this trend will continue. My extremely rough guess is that I currently approve of 4% of language features. ;)

JavaScript's core: `function f(x) { return x; }`, `(`, `)`, `[`, `]`, `{`, `}`, `'abc'`, `123`. Almost everything else can be implemented with these primitives.

Reply:
I agree with the general sentiment, but I guess we differ on where we draw the line.

I don't view: `{ foo: 42, bar: 10 }` as a primitive of the language. Instead I view it as a sugar for `{ foo: 42, ...({ bar: 10, ...({}) }) }`, i.e. I see: `{} :: Object k v o` and `{ [k]: v, ...o } :: Object k v o` as the actual primitives of the language (similar for `[]` and `[x, ...xs]`).

As a result I think it's fine to use language features that pattern match and pun over `[]` and `{}` literals.

Reply:
To be clear, the decision to target ECMAScript 5.1 is practical rather than philosophical. Were newer language features supported everywhere I would absolutely adopt a few. :)